packages/sdk-js/src/handlers/auth.ts

```diff
@@ -42,45 +48,74 @@ export class AuthHandler {
 	}
 
 	async login(credentials: LoginCredentials) {
+		clearTimeout(this.autoRefreshTimeout);
 		const response = await this.axios.post('/auth/login', { ...credentials, mode: this.mode });
 
-		this.token = response.data.data.access_token;
+		const data = response.data.data;
+		this.token = data.access_token;
+		this.expiresAt = Date.now() + data.expires;
+
+		await this.storage.setItem('directus_access_token', this.token);
+		await this.storage.setItem('directus_access_token_expires', this.expiresAt);
 
 		if (this.mode === 'json') {
-			await this.storage.setItem('directus_refresh_token', response.data.data.refresh_token);
+			await this.storage.setItem('directus_refresh_token', data.refresh_token);
 		}
-
 		if (this.autoRefresh) {
-			setTimeout(() => this.refresh(), response.data.data.expires - 10000);
+			this.refresh();
 		}
 
 		return response.data;
 	}
 
+	/**
+	 * Refresh access token 10 seconds before expiration
+	 */
 	async refresh() {
+		clearTimeout(this.autoRefreshTimeout);
+
+		this.expiresAt = await this.storage.getItem('directus_access_token_expires');
+		if (!this.expiresAt) return;
+
+		if (Date.now() + 10000 < this.expiresAt) {
+			this.autoRefreshTimeout = setTimeout(
+				() => this.refresh(),
+				this.expiresAt - Date.now() - 10000
+			);
+			return;
+		}
+
 		const payload: Record<string, any> = { mode: this.mode };
 
 		if (this.mode === 'json') {
 			const refreshToken = await this.storage.getItem('directus_refresh_token');
 			payload['refresh_token'] = refreshToken;
 		}
 
-		const response = await this.axios.post('/auth/refresh', payload);
+		const response = await this.axios.post('/auth/refresh', payload, {
+			// If access_token is expired, it will throw 401, so we remove it from headers
+			transformRequest: (data, headers) => {
+				delete headers.Authorization;
+				return data;
+			},
```

Comment:
We should remove it from the axios headers by setting `this.token` to null. That should set the default header in axios

Reply:
Should we remove it?
If this happens:
sdk sends auto refresh request without awaiting and deletes this.token
user sends request
sdk finishes auto refresh request
user will send request without access token and will get 401, without knowing the reason.

This way we:
sdk sends auto refresh without awaiting, and leaves access token
user sends request with valid access token
sdk finishes auto refresh request
user receives data as normal

Reply:
I meant that we should try to keep all axios headers defaulting within that `token` getter/setter instead of deleting those elsewhere, otherwise it'll get a little difficult to track where what is happening with that token later on 🙂

Reply:
Something like this? This will delete access token if there is less then 1s remaining. 🙂
```ts
if (this.expiresAt < Date.now() + 1000) {
	this.token = null;
}
const response = await this.axios.post('/auth/refresh', payload);


```

Reply:
lgtm